sklearn/preprocessing/_encoders.py

```diff
@@ -1446,6 +1481,103 @@ def __init__(
+                raise ValueError(
+                    f"{self.__class__.__name__} was not fitted on data "
+                    "with feature names. Pass max_categories as an integer "
+                    "array instead."
+                )
+            unexpected_feature_names = list(
+                set(self.max_categories) - set(self.feature_names_in_)
+            )
+            unexpected_feature_names.sort()  # deterministic error message
+            n_unexpeced = len(unexpected_feature_names)
+            if unexpected_feature_names:
+                if len(unexpected_feature_names) > 5:
+                    unexpected_feature_names = unexpected_feature_names[:5]
+                    unexpected_feature_names.append("...")
+                raise ValueError(
+                    f"max_categories contains {n_unexpeced} unexpected feature "
+                    f"names: {unexpected_feature_names}."
+                )
+            for feature_idx, feature_name in enumerate(self.feature_names_in_):
+                if feature_name in self.max_categories:
+                    max_count = self.max_categories[feature_name]
+                    if not (isinstance(max_count, Integral) and max_count >= 1):
+                        raise ValueError(
+                            f"max_categories['{feature_name}'] must be an "
+                            f"integer at least 1. Got {max_count!r}."
+                        )
+                    max_categories[feature_idx] = max_count
+        else:
+            max_categories = np.asarray(max_categories)
+            if max_categories.shape[0] != self.n_features_in_:
+                raise ValueError(
+                    f"max_categories has shape {max_categories.shape} but the "
+                    f"input data X has {self.n_features_in_} features."
+                )
+
+            unexpected_max_counts = set(
+                max_count
+                for max_count in max_categories
+                if not (
+                    max_count is None
+                    or (isinstance(max_count, Integral) and max_count >= 1)
+                )
+            )
+            if len(unexpected_max_counts):
+                raise ValueError(
+                    "max_categories must be an array-like of None or integers "
+                    "at least 1. Observed "
+                    f"values: {list(unexpected_max_counts)}."
+                )
+
+        self.max_categories = max_categories
+
+    def _max_categories_enable_infrequent(self, max_categories):
+        """
+        This function checks whether the value of max_categories
+        enables infrequent categories.
+        """
+        if max_categories is None:
+            return False
+        elif isinstance(max_categories, Integral):
+            return max_categories >= 1
+        else:
+            return any(max_categories)
+
+    def _has_infrequent_categories(self, n_current_features, col_idx):
```

Comment:
Yes, `col_idx` really should be renamed as `feature_idx` as it describes the index of the feature of the `X` that was passed in. I kept it as `col_idx` since the original code uses it and I want to maintain consistency until someone more well-versed in the codebase provides an opinion.